packages/jss/src/index.d.ts

```diff
@@ -1,35 +1,156 @@
-// Type definitions for jss 9.5
-// Project: https://github.com/cssinjs/jss#readme
-// Definitions by: Brenton Simpson <https://github.com/appsforartists>
-//                 Oleg Slobodskoi <https://github.com/kof>
-//                 Thomas Crockett <https://github.com/pelotom>
-//                 Sebastian Silbermann <https://github.com/eps1lon>
-// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-// TypeScript Version: 2.3
+export type Style = object;
+export type Styles<Name extends string = string> = Record<Name, Style>;
+export type Classes<Name extends string = string> = Record<Name, string>;
+export type Keyframes<Name extends string = string> = Record<Name, string>;
 
-import { Style } from './css';
+export type GenerateId = (rule: Rule, sheet?: StyleSheet<string>) => string
 
-export type Styles<Name extends string = any> = Record<Name, Style>;
-export type Classes<Name extends string = any> = Record<Name, string>;
+export type JssValue =
+  | string
+  | number
+  | Array<string | number | Array<string | number> | '!important'>
+  | null
+  | false
+
+export type CreateGenerateId = () => GenerateId
+
+export type InsertionPoint = string | HTMLElement
+
+export interface UpdateOptions {
+  process?: boolean
+  force?: boolean
+}
 
 export interface ToCssOptions {
   indent?: number;
+  allowEmpty?: boolean
+}
+
+interface RuleListOptions {
+  classes: Classes
+  generateClassName: GenerateId
+  Renderer: Renderer
+  jss: Jss
+  sheet: StyleSheet
+  parent: ContainerRule | StyleSheet
+}
+
+declare class RuleList {
+  constructor(options: RuleListOptions)
+  add(name: string, decl: Style, options?: RuleOptions): Rule
+  get(name: string): Rule
+  remove(rule: Rule): void
+  indexOf(rule: Rule): number
+  process(): void
+  register(rule: Rule, className?: string): void
+  unregister(rule: Rule): void
+  update(name: string, data: {}): void
+  update(data: {}): void
+  toString(options?: ToCssOptions): string
+}
+
+interface RuleOptions {
+  selector?: string
+  sheet?: StyleSheet
+  index?: number
+  parent?: ContainerRule | StyleSheet
+  classes: Classes
+  jss: Jss
+  generateId: GenerateId
+  Renderer: Renderer
+}
+
+declare class BaseRule {
+  type: string
+  key: string
+  isProcessed: boolean
+  // eslint-disable-next-line no-use-before-define
+  options: RuleOptions
+  constructor(key: string, style: Style, options: RuleOptions)
+  toString(options?: ToCssOptions): string
+}
+
+interface ContainerRule extends BaseRule {
+  rules: RuleList
+}
+
+export interface Plugin {
+  onCreateRule?(name: string, decl: Style, options: RuleOptions): Rule
+  onProcessRule?(rule: Rule, sheet?: StyleSheet): void
+  onProcessStyle?(style: Style, rule: Rule, sheet?: StyleSheet): Style
+  onProcessSheet?(sheet?: StyleSheet): void
```

Comment:
Do `onProcessSheet` or `onProcessRule` ever get called without `sheet`?

Reply:
`onProcessSheet` shouldn't, but `onProcessRule` can be called without a sheet, so this is very rare.